File: src/cpp/predict.cpp

#define FMT_HEADER_ONLY

#include <chrono>
#include <deque>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <mutex>
#include <stdexcept>
#include <string>
#include <unordered_map>

#include <ATen/ops/nonzero.h>
#include <Magick++.h>
#include <fmt/core.h>
#include <getopt.h>
#include <nlohmann/json.hpp>
#include <nlohmann/json_fwd.hpp>
#include <opencv2/core/cvstd.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/opencv.hpp>
#include <spdlog/spdlog.h>
#include <zmq.hpp>

#include "model_utils.h"
#include "rest/oatpp_entry.h"
#include "utils.h"

using namespace std;
using json = nlohmann::json;

volatile sig_atomic_t ev_flag = 0;
std::atomic<uint32_t> inference_interval_ms = 60000;
vector<string> model_ids;
json settings;
mutex image_queue_mtx, ext_program_mtx, swagger_mtx, models_mtx, model_ids_mtx;
deque<vector<char>> image_queue;
constexpr ssize_t gif_frame_count = 36;
constexpr ssize_t inference_batch_size = 24;
constexpr ssize_t pre_detection_size = 4;
constexpr ssize_t image_queue_min_len =
    pre_detection_size + inference_batch_size + gif_frame_count;
constexpr ssize_t image_queue_max_len = image_queue_min_len * 4;
string cuda_device_string = "cuda:0";
string torch_script_serialization;
vector<torch::jit::script::Module> models;
std::unordered_map<uint32_t, PercentileTracker<float>> pt_dict;

void print_usage(string binary_name) {

  cerr << "Usage: " << binary_name << " [OPTION]\n\n";

  cerr << "Options:\n"
       << "  --help,        -h        Display this help and exit\n"
       << "  --config-path, -c        JSON configuration file path" << endl;
}

void parse_arguments(int argc, char **argv, string &config_path) {
  static struct option long_options[] = {
      {"config-path", required_argument, 0, 'c'},
      {"help", optional_argument, 0, 'h'},
      {0, 0, 0, 0}};

  int opt, option_index = 0;

  while ((opt = getopt_long(argc, argv, "c:h", long_options, &option_index)) !=
         -1) {
    switch (opt) {
    case 'c':
      if (optarg != NULL) {
        config_path = string(optarg);
      }
      break;
    default:
      print_usage(argv[0]);
      exit(EXIT_FAILURE);
    }
  }
  if (config_path.empty()) {
    print_usage(argv[0]);
    exit(EXIT_FAILURE);
  }
}

void zeromq_ev_loop() {
  spdlog::info("zeromq_ev_loop() started");

  zmq::context_t context(1);
  zmq::socket_t subscriber(context, ZMQ_SUB);
  string zmq_address = settings.value("/inference/zeromq_address"_json_pointer,
                                      "tcp://127.0.0.1:4240");
  subscriber.connect(zmq_address);
  spdlog::info("ZeroMQ client connected to {}", zmq_address);
  int timeout = 5000;
  subscriber.setsockopt(ZMQ_RCVTIMEO, &timeout, sizeof(timeout));
  subscriber.setsockopt(ZMQ_SUBSCRIBE, "", 0);
  while (!ev_flag) {
    zmq::message_t message;
    try {
      subscriber.recv(&message);
    } catch (const zmq::error_t &e) {
      if (e.num() == EAGAIN) {
        spdlog::warn("subscriber.recv() timed out, will retry");
        continue;
      } else {
        spdlog::error("subscriber.recv() throws exception: {}", e.what());
        continue;
      }
    }
    auto t = vector<char>(message.size());
    if (message.size() == 0) {
      spdlog::warn("ZeroMQ received empty message");
      continue;
    }
    memcpy(t.data(), message.data(), t.size());
    {
      lock_guard<mutex> lock(image_queue_mtx);
      image_queue.push_back(t);
      while (image_queue.size() > image_queue_max_len) {
        image_queue.pop_front();
      }
    }
  }
  subscriber.close();
  spdlog::info("ZeroMQ connection closed");
  spdlog::info("zeromq_ev_loop() exited gracefully");
}

void execute_external_program_async() {
  auto f = [](string cmd) {
    if (ext_program_mtx.try_lock()) {
      spdlog::info("Calling external program: {}", cmd);
      system(cmd.c_str());
      ext_program_mtx.unlock();
    } else {
      spdlog::warn("ext_program_mtx is locked already, meaning that another {} "
                   "instance is already running",
                   cmd);
    }
  };
  thread th_exec(f, settings["inference"]["on_detected"]["external_program_cpp"]
                        .get<string>());
  th_exec.detach();
}

bool handle_inference_results(vector<at::Tensor> &outputs, at::Tensor &output,
                              const vector<vector<char>> &jpegs) {
  ostringstream oss_raw_result, oss_avg_result;

  for (size_t i = 0; i < outputs.size(); ++i) {
    oss_raw_result << outputs[i];
  }
  oss_avg_result << output;
  at::Tensor y_pred = torch::argmax(output, 1);

  auto nonzero_y_preds_idx = torch::nonzero(y_pred);
  if (nonzero_y_preds_idx.sizes()[0] == 0) {
    return false;
  }
  spdlog::warn("Target detected at {}-th frame in a batch of {} frames",
               nonzero_y_preds_idx[0].item<int>(), inference_batch_size, 0);
  spdlog::info("y_pred:              {}",
               tensor_to_string_like_pytorch(y_pred, 0, y_pred.sizes()[0]));
  spdlog::info("nonzero_y_preds_idx: {}",
               tensor_to_string_like_pytorch(nonzero_y_preds_idx, 0,
                                             nonzero_y_preds_idx.sizes()[0]));
  spdlog::info("Raw results are:\n{}", oss_raw_result.str());
  spdlog::info("and arithmetic average of raw results is:\n{}",
               oss_avg_result.str());
  spdlog::info("Build GIF animation");
  vector<Magick::Image> frames;
  spdlog::info("jpegs[{}: {}] of {} images will be used to build the GIF "
               "animation",
               nonzero_y_preds_idx[0].item<int>(),
               nonzero_y_preds_idx[0].item<int>() + gif_frame_count,
               jpegs.size());
  auto gif_width =
      settings.value("/inference/on_detected/gif_size/width"_json_pointer,
                     target_img_size.width);
  auto gif_height =
      settings.value("/inference/on_detected/gif_size/height"_json_pointer,
                     target_img_size.height);
  // nonzero_y_preds_idx[0] stores the first non-zero item
  // index in y_pred. Note that y_pred[0] is NOT the result
  // of jpegs[0], but jpegs[pre_detection_size]
  for (int i = 0; i < gif_frame_count; ++i) {
    // One needs to think for a while to understand the
    // offset between jpegs_idx and y_pred's index--their gap
    // is exactly inference_batch_size, which is implied in
    // the line: cv::imdecode(jpegs[pre_detection_size + i],
    // cv::IMREAD_COLOR));
    auto jpegs_idx = nonzero_y_preds_idx[0].item<int>() + i;
    assert(jpegs_idx < jpegs.size());
    frames.emplace_back(
        Magick::Blob(jpegs[jpegs_idx].data(), jpegs[jpegs_idx].size()));
    frames.back().animationDelay(10); // 100 milliseconds (10 * 1/100th)
    frames.back().resize(Magick::Geometry(gif_width, gif_height));
  }
  string gif_path = settings.value(
      "/inference/on_detected/gif_path"_json_pointer, "/tmp/detected.gif");
  spdlog::info("Saving GIF file (size: {}x{}) to [{}]", gif_width, gif_height,
               gif_path);
  Magick::writeImages(frames.begin(), frames.end(), gif_path);
  execute_external_program_async();
  filesystem::path jpg_dir = settings.value(
      "/inference/on_detected/jpegs_directory"_json_pointer, "/");
  spdlog::info("Saving positive images as JPEG files to [{}]",
               jpg_dir.native());
  for (int i = 0; i < nonzero_y_preds_idx.sizes()[0]; ++i) {
    auto jpegs_idx = nonzero_y_preds_idx[i].item<int>() + pre_detection_size;
    assert(jpegs_idx < jpegs.size());
    filesystem::path jpg_path = jpg_dir / (getCurrentDateTimeString() + ".jpg");
    ofstream outFile(jpg_path, ios::binary);
    if (!outFile) {
      spdlog::error("Failed to open the file [{}]", jpg_path.native());
    } else {
      outFile.write(jpegs[jpegs_idx].data(), jpegs[jpegs_idx].size());
      outFile.close();
    }
  }
  return true;
}

pair<vector<at::Tensor>, at::Tensor>
infer_images(vector<torch::jit::script::Module> &models,
             const vector<vector<char>> &jpegs) {
  auto start = chrono::high_resolution_clock::now();
  vector<torch::Tensor> images_tensors_vec(inference_batch_size);
  torch::Tensor images_tensor;
  vector<torch::jit::IValue> input(1);
  for (int i = 0; i < inference_batch_size; ++i) {
    // images_mats[i] = cv::imdecode(received_jpgs[i], cv::IMREAD_COLOR);
    images_tensors_vec[i] = cv_mat_to_tensor(
        cv::imdecode(jpegs[pre_detection_size + i], cv::IMREAD_COLOR));
  }
  images_tensor = torch::stack(images_tensors_vec);

  input[0] = images_tensor.to(cuda_device_string);

  // images_tensor.sizes()[0] stores number of images
  at::Tensor output =
      torch::zeros({images_tensor.sizes()[0], NUM_OUTPUT_CLASSES});
  output = output.to(cuda_device_string);
  vector<at::Tensor> outputs(models.size());
  {
    std::lock_guard<std::mutex> lock(models_mtx);
    for (size_t i = 0; i < models.size(); ++i) {
      outputs[i] = models[i].forward(input).toTensor();
      output += outputs[i];
    }
  }
  auto end = chrono::high_resolution_clock::now();
  {
    lock_guard<mutex> lock(swagger_mtx);
    auto [it, success] = pt_dict.try_emplace(inference_interval_ms,
                                             PercentileTracker<float>(10000));
    pt_dict.at(inference_interval_ms)
        .addNumber(
            (float)chrono::duration_cast<chrono::milliseconds>(end - start)
                .count() /
            inference_batch_size);
  }
  return make_pair(outputs, output);
}

void interruptible_sleep(const size_t sleep_ms) {
  const size_t interruptible_sleep_ms = 1000;
  if (sleep_ms <= interruptible_sleep_ms) {
    this_thread::sleep_for(chrono::milliseconds(sleep_ms));
  } else {
    size_t slept_ms = 0;
    while (slept_ms < sleep_ms && !ev_flag) {
      slept_ms += interruptible_sleep_ms;
      this_thread::sleep_for(chrono::milliseconds(interruptible_sleep_ms));
    }
  }
}

void inference_ev_loop() {
  int delay_ms = 10000;
  spdlog::info("inference_ev_loop() started, waiting for {}ms before start",
               delay_ms);
  // Otherwise we won't have any images of detected object in the GIF
  assert(pre_detection_size < gif_frame_count);
  assert(gif_frame_count > 0);
  assert(inference_batch_size > 0);
  // Initial wait, just to spread out the stress at the beginning of the run
  interruptible_sleep(delay_ms);
  try {
    {
      std::scoped_lock lck{models_mtx, model_ids_mtx};
      models = load_models(model_ids);
    }

  } catch (const c10::Error &e) {
    spdlog::error("Error loading the model: {}", e.what());
    ev_flag = 1;
    spdlog::critical("The program must exit now");
  }

  Magick::InitializeMagick(nullptr);
  vector<vector<char>> received_jpgs(image_queue_min_len);
  vector<cv::Mat> images_mats(inference_batch_size);

  while (!ev_flag) {
    interruptible_sleep(inference_interval_ms);
    {
      lock_guard<mutex> lock(image_queue_mtx);
      if (image_queue.size() < image_queue_min_len) {
        spdlog::warn("image_queue.size() == {} < image_queue_min_len({}), "
                     "waiting for more images before inference can run",
                     image_queue.size(), image_queue_min_len);
        continue;
      }
      for (int i = 0; i < image_queue_min_len; ++i) {
        received_jpgs[i] = image_queue[i];
      }
      for (int i = 0; i < inference_batch_size; ++i) {
        image_queue.pop_front();
      }
    }
    auto rv = infer_images(models, received_jpgs);
    if (handle_inference_results(rv.first, rv.second, received_jpgs)) {
      const size_t cooldown_ms =
          settings.value("/inference/on_detected/cooldown_sec"_json_pointer,
                         120) *
          1000;
      spdlog::info("inference_ev_loop() will sleep for {}ms after detection",
                   cooldown_ms);
      interruptible_sleep(cooldown_ms);
    }
  }
  spdlog::info("inference_ev_loop() exited gracefully");
}

int main(int argc, char **argv) {
  install_signal_handler();
  string config_path;
  parse_arguments(argc, argv, config_path);
  spdlog::set_pattern("%Y-%m-%d %T.%e | %7l | %5t | %v");
  spdlog::info("Predict daemon started");
  spdlog::info("Loading configurations from {}:", config_path);
  ifstream f(config_path);
  settings = json::parse(f);
  spdlog::info("{}", settings.dump(2));
  model_ids = settings.value("/inference/initial_model_ids"_json_pointer,
                             vector<string>{"0"});
  torch_script_serialization = settings.value(
      "/model/torch_script_serialization"_json_pointer, string(""));

  initialize_rest_api(
      settings.value("/inference/swagger/host"_json_pointer, "127.0.0.1"),
      settings.value("/inference/swagger/port"_json_pointer, 8000),
      settings.value("/inference/swagger/advertised_host"_json_pointer,
                     "http://127.0.0.1:8000"));
  inference_interval_ms =
      settings.value("/inference/initial_interval_ms"_json_pointer, 60000);
  thread thread_zeromq(zeromq_ev_loop);
  thread thread_inference(inference_ev_loop);
  if (thread_zeromq.joinable()) {
    thread_zeromq.join();
  }
  if (thread_inference.joinable()) {
    thread_inference.join();
  }
  finalize_rest_api();
  spdlog::info("Predict daemon exiting");
  return 0;
}

File: src/py/predict.py
from typing import Any, Dict
from flask import Flask, request, Response
from threading import Thread, Lock
from PIL import Image
from torch.utils.data import DataLoader, Dataset
from typing import List, Any

import datetime as dt
import helper
import io
import logging
import json
import model
import os
import time
import signal
import sqlite3
import subprocess
import torch
import torchvision
import waitress
import sys
import zmq


app = Flask(__name__)
curr_dir = os.path.dirname(os.path.realpath(__file__))
prediction_interval = 600
ev_flag = True
image_queue_mutex = Lock()
image_queue: List[Any] = []
image_queue_min_len = 64
image_queue_max_len = image_queue_min_len * 2
image_context_start = -3
image_context_end = 12
db_path = os.path.join(curr_dir, 'predict.sqlite')


def signal_handler(signum: int, frame: Any) -> None:
    global ev_flag
    print('Signal handler called with signal', signum)
    ev_flag = False
    signal.signal(signum, signal.SIG_DFL)


@app.route("/")
def set_prediction_interval() -> Response:
    global prediction_interval
    try:
        prediction_interval = float(request.args.get('prediction_interval', '1'))
        logging.info(f'prediction_interval changed to {prediction_interval}')
    except Exception as ex:
        logging.exception('Unabled to set new prediction_interval')
        return Response(f"{ex}", status=400)
    return Response(f'prediction_interval: {prediction_interval} sec', status=200)


def prepare_database() -> None:

    conn = sqlite3.connect(db_path)
    cur = conn.cursor()

    cur.execute('''
        CREATE TABLE IF NOT EXISTS inference_records (
            id INTEGER PRIMARY KEY,
            timestamp TEXT,
            model_output TEXT,
            prediction INTEGER,
            elapsed_time_ms REAL
    )''')
    conn.commit()

    cutoff_date = dt.datetime.now() - dt.timedelta(days=15)

    cur.execute(
        "DELETE FROM inference_records WHERE timestamp < ?", (cutoff_date,)
    )
    conn.commit()


def zeromq_thread() -> None:

    context = zmq.Context()
    url = "tcp://127.0.0.1:4241"
    #  Socket to talk to server
    logging.info(f"Connecting to publisher endpoint [{url}]")
    socket = context.socket(zmq.SUB)
    socket.connect(url)
    socket.setsockopt(zmq.SUBSCRIBE, b'')
    logging.info("Connected to endpoint")

    while ev_flag:
        #  Get the reply.
        message = socket.recv()
        image_queue_mutex.acquire()
        while len(image_queue) >= image_queue_max_len:
            image_queue.pop(0)
        image_queue.append(message)
        image_queue_mutex.release()

    logging.info('zeromq_thread() exited gracefully')


def insert_prediction_to_db(cur: sqlite3.Cursor, model_output: str,
                            prediction: int, elapsed_time_ms: float) -> None:
    sql = """
        INSERT INTO inference_records(
            timestamp, model_output, prediction, elapsed_time_ms
        )
        VALUES (?, ?, ?, ?)
    """
    cur.execute(sql, (dt.datetime.now().isoformat(), model_output,
                      prediction, elapsed_time_ms))


def initialize_logger() -> None:
    root = logging.getLogger()
    root.setLevel(logging.DEBUG)

    handler = logging.StreamHandler(sys.stdout)
    handler.setLevel(logging.DEBUG)
    formatter = logging.Formatter(
        '%(asctime)s.%(msecs)03d | %(name)8s | %(levelname)7s | %(message)s',
    )
    handler.setFormatter(formatter)
    root.addHandler(handler)


class CustomDataset(Dataset):
    def __init__(
        self, images, transform: torchvision.transforms.Compose = None
    ) -> None:
        self.images = images
        self.transform = transform

    def __len__(self) -> int:
        return len(self.images)

    def __getitem__(self, idx: int) -> Image:
        image = self.images[idx]
        if self.transform:
            image = self.transform(image)
        return image


def prediction_thread() -> None:

    logging.info('prediction_thread() started')
    global prediction_interval
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    settings: Dict[str, Any]
    with open(os.path.join(curr_dir, '..', '..', 'config.json')) as j:
        settings = json.load(j)

    conn = sqlite3.connect(db_path)
    cur = conn.cursor()
    model_ids = ['0', '1', '2']
    v16mms = []
    for i in range(len(model_ids)):
        v16mms.append(model.VGG16MinusMinus(2))
        v16mms[i].to(device)
        model_path = settings['model']['parameters'].replace(
            r'{id}', model_ids[i]
        )
        logging.info(f'Loading weights to model from: {model_path}')
        v16mms[i].load_state_dict(torch.load(model_path))
        total_params = sum(p.numel() for p in v16mms[i].parameters())
        logging.info(f"Weights loaded, number of parameters: {total_params:,}")
        v16mms[i].eval()

    DATASET_SIZE = 16
    assert DATASET_SIZE + image_context_start > 0
    # assert image_context_end - image_context_start == DATASET_SIZE
    iter_count = 0
    while ev_flag:
        iter_count += 1
        image_queue_mutex.acquire()
        logging.info("Iterating prediction loop "
                     f"(len(image_queue): {len(image_queue)})...")
        if len(image_queue) < image_queue_min_len:
            logging.warning(
                f'len(image_queue): {len(image_queue)}, '
                'waiting for more frames...'
            )
            image_queue_mutex.release()
            count = 0
            while count < 5 and ev_flag:
                count += 1
                time.sleep(1)
            continue

        start_time = time.time()

        # Create an instance of your custom dataset class
        dataset = CustomDataset([
            Image.open(io.BytesIO(image_queue[DATASET_SIZE + 0])),
            Image.open(io.BytesIO(image_queue[DATASET_SIZE + 1])),
            Image.open(io.BytesIO(image_queue[DATASET_SIZE + 2])),
            Image.open(io.BytesIO(image_queue[DATASET_SIZE + 3])),
            Image.open(io.BytesIO(image_queue[DATASET_SIZE + 4])),
            Image.open(io.BytesIO(image_queue[DATASET_SIZE + 5])),
            Image.open(io.BytesIO(image_queue[DATASET_SIZE + 6])),
            Image.open(io.BytesIO(image_queue[DATASET_SIZE + 7])),
            Image.open(io.BytesIO(image_queue[DATASET_SIZE + 8])),
            Image.open(io.BytesIO(image_queue[DATASET_SIZE + 9])),
            Image.open(io.BytesIO(image_queue[DATASET_SIZE + 10])),
            Image.open(io.BytesIO(image_queue[DATASET_SIZE + 11])),
            Image.open(io.BytesIO(image_queue[DATASET_SIZE + 12])),
            Image.open(io.BytesIO(image_queue[DATASET_SIZE + 13])),
            Image.open(io.BytesIO(image_queue[DATASET_SIZE + 14])),
            Image.open(io.BytesIO(image_queue[DATASET_SIZE + 15]))
        ], transform=helper.test_transforms)
        assert DATASET_SIZE == len(dataset.images)
        dataloader = DataLoader(
            dataset, batch_size=DATASET_SIZE, shuffle=False, num_workers=4
        )

        # Classify the image using the model
        torch.set_grad_enabled(False)
        outputs: List[torch.Tensor] = []
        output = torch.zeros(
            [DATASET_SIZE, v16mms[0].num_classes], dtype=torch.float32
        )
        output = output.to(device)
        # Per current, batch_size is always equal to DATASET_SIZE
        # so there will always be only one batch
        for batch in dataloader:
            batch = batch.to(device)
            for i in range(len(v16mms)):
                outputs.append(v16mms[i](batch))
                output += outputs[i]
            _, pred_tensor = torch.max(output.data, 1)
        torch.set_grad_enabled(True)
        elapsed_time = time.time() - start_time
        for i in range(DATASET_SIZE):
            insert_prediction_to_db(cur, str(output[i]), int(pred_tensor[i]),
                                    round(elapsed_time * 1000.0, 1))

        if iter_count > 15:
            # commit() could be a very expensive operation
            # profiling shows it takes 1+ sec to complete
            conn.commit()
            logging.info('SQLite commit()ed')
            iter_count = 0

        nonzero_preds = torch.nonzero(pred_tensor)
        if len(nonzero_preds) > 0:
            logging.warning(
                f'Target detected at {nonzero_preds[0].item()}-th frame in '
                f'a batch of {DATASET_SIZE} frames '
                f'({nonzero_preds[0].item() + DATASET_SIZE}-th frame in the '
                f'queue of {len(image_queue)} frames), '
                'preparing context frames')
            logging.warning(f'The entire output is: {pred_tensor}')
            logging.warning(f'Raw results from {len(outputs)} models are:')
            for i in range(len(outputs)):
                output_str = ''
                for j in range(len(nonzero_preds)):
                    output_str += f'{nonzero_preds[j]}]: ' + str(outputs[i][nonzero_preds[j]]) + '\n'
                logging.warning(f'\n[{output_str}')
            output_str = ''
            for i in range(len(nonzero_preds)):
                output_str += f'{nonzero_preds[i]}]: ' + str(output[nonzero_preds[i]]) + '\n'
            logging.warning(f'and arithmetic average of raw results is:\n{output_str}')
            for i in range(image_context_end - image_context_start):
                temp_img_path = f'/tmp/frame{i}.jpg'
                with open(temp_img_path, "wb") as binary_file:
                    image_idx = int(
                        DATASET_SIZE + nonzero_preds[0].item() +
                        image_context_start + i * 2
                    )
                    logging.info(f'Writing {image_idx}-th frame from the queue '
                                 f'to path [{temp_img_path}]')
                    binary_file.write(image_queue[image_idx])

            for i in range(DATASET_SIZE):
                image_queue.pop(0)
            image_queue_mutex.release()
            logging.info(
                f'Calling program [{settings["inference"]["on_detected"]}]...'
            )
            result = subprocess.run(
                settings['inference']['on_detected']['external_program_py'],
                stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True
            )
            logging.info(f'stdout: {result.stdout}')
            logging.info(f'stderr: {result.stderr}')

            cooldown_sec = 150
            logging.info(f'Entered cooldown period ({cooldown_sec} sec)')
            count = 0
            while count < cooldown_sec * 2 and ev_flag:
                count += 1
                time.sleep(0.5)
        else:

            for i in range(DATASET_SIZE):
                image_queue.pop(0)
            image_queue_mutex.release()

        count = 0
        while count < prediction_interval * 10.0 and ev_flag:
            count += 1
            time.sleep(0.1)
    conn.commit()
    logging.info('SQLite commit()ed')
    logging.info('prediction_thread() exited gracefully')
    conn.close()


def main() -> None:
    initialize_logger()
    logging.info('predict.py started')
    prepare_database()
    signal.signal(signal.SIGINT, signal_handler)
    th_pred = Thread(target=prediction_thread)
    th_pred.start()
    th_zmq = Thread(target=zeromq_thread)
    th_zmq.start()
    waitress.serve(app, host='127.0.0.1', port='4386')
    logging.info('main() exited gracefully')


if __name__ == '__main__':
    main()